### PR TITLE
dsp/window: use half offset to exclude flanking zeros

### LINKED ISSUE #1240: dsp/window: revisit window function definitions' use of N-1 vs N
See https://github.com/gonum/gonum/pull/1230#pullrequestreview-360730303

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Closes #1240.

### CODE REVIEW

**Comment by kortschak** [comment]
cc @goofinator

File: dsp/window/doc.go [hunk 1]
```diff
@@ -40,4 +40,9 @@
 //
 // The ɣ_max parameter is the maximum level of the side lobes of the
 // normalized spectrum, in decibels.
+//
+// Window interval
+//
+// The window intervals used in the window function are offset by a half
```
Comment (sbinet) [inline dsp/window/doc.go]:
```suggestion
// The window intervals used in the window functions are offset by a half
```

File: dsp/window/window_test.go [hunk 2]
```diff
@@ -5,82 +5,111 @@
...
-		name: "Lanczos", fn: Lanczos, winLen: 20,
-		want: []float64{0.000000, 0.115514, 0.247646, 0.389468, 0.532984, 0.669692, 0.791213, 0.889915, 0.959492, 0.995450,
-			0.995450, 0.959492, 0.889915, 0.791213, 0.669692, 0.532984, 0.389468, 0.247646, 0.115514, 0.000000},
+		name: "Lanczos", fn: Lanczos, fnCmplx: LanczosComplex, winLen: 20,
+		want: []float64{
+			0.052415, 0.170011, 0.300105, 0.436333, 0.57162, 0.698647, 0.810332, 0.900316, 0.963398, 0.995893,
+			0.995893, 0.963398, 0.900316, 0.810332, 0.698647, 0.57162, 0.436333, 0.300105, 0.170011, 0.052415,
+		},
 	},
 	// This case tests Lanczos for a NaN condition. The Lanczos NaN condition is k=(N-1)/2, that is when N is odd.
```
Comment (goofinator) [inline dsp/window/window_test.go]: // This case tests Lanczos for a NaN condition. The Lanczos NaN condition is k=(N-2)/2, 
	// that is when N is even and N>=2.
So, test with winLen=21 not checks NaN condition now.

File: dsp/window/window.go [hunk 3]
```diff
@@ -30,14 +30,14 @@ func Rectangular(seq []float64) []float64 {
 // Sine window is a high-resolution window.
 //
 // The sequence weights are
-//  w[k] = sin(π*k/(N-1)),
+//  w[k] = sin(π*k/N),
```
Comment (vladimir-ch) [inline dsp/window/window.go]: This description (and others) misses the `+ 1/2` offset.

File: dsp/window/window_test.go [hunk 4]
```diff
@@ -5,82 +5,111 @@
...
 
 var windowTests = []struct {
-	name   string
-	fn     func([]float64) []float64
-	want   []float64
-	winLen int
+	name    string
+	fn      func([]float64) []float64
+	fnCmplx func([]complex128) []complex128
+	winLen  int
```
Comment (vladimir-ch) [inline dsp/window/window_test.go]: I think `winLen` can be taken from the length of `want`.

File: dsp/window/window_test.go [hunk 5]
```diff
@@ -119,7 +154,7 @@ func TestWindows(t *testing.T) {
 			dst := test.fn(src)
 
 			if !floats.EqualApprox(dst, test.want, tol) {
```
Comment (vladimir-ch) [inline dsp/window/window_test.go]: It's not so critical here but in general when comparing two slices/vectors/matrices I prefer to get a measure of their distance and then compare this value to a tolerance. It's then easy to log the distance and see if a failure is due to too strict tolerance or an actual bug.

File: dsp/window/window_test.go [hunk 6]
```diff
@@ -134,15 +169,72 @@ func TestGausWindows(t *testing.T) {
...
+	const tol = 1e-6
+
+	src := make([]complex128, 20)
+	for i := range src {
+		src[i] = complex(1, 1)
+	}
+
+	for _, test := range gausWindowTests {
+		t.Run(fmt.Sprintf("%sComplex (sigma=%.1f)", test.name, test.sigma), func(t *testing.T) {
+			// Copy the input since we are mutating it.
```
[the remark this answers is not in the document]
Comment (kortschak) [reply dsp/window/window_test.go]: Done.

File: dsp/window/window_test.go [hunk 7]
```diff
@@ -111,6 +145,7 @@ func TestWindows(t *testing.T) {
 
 	for _, test := range windowTests {
 		t.Run(test.name, func(t *testing.T) {
+			// Copy the input since we are mutating it.
```
[the remark this answers is not in the document]
Comment (kortschak) [reply dsp/window/window_test.go]: Done.

File: dsp/window/window_test.go (hunk 4 above)
Comment (kortschak) [reply dsp/window/window_test.go]: Done.

File: dsp/window/doc.go (hunk 1 above)
Comment (kortschak) [reply dsp/window/doc.go]: Done.

File: dsp/window/window.go (hunk 3 above)
Comment (kortschak) [reply dsp/window/window.go]: I had intended to show the general case and implement the specific case, but I'll fix this.

File: dsp/window/window_test.go (hunk 5 above)
Comment (kortschak) [reply dsp/window/window_test.go]: Yeah. The slices here are short and we show them, so I'm OK with doing it this way.

File: dsp/window/window.go (hunk 3 above)
Comment (vladimir-ch) [reply dsp/window/window.go]: If you did it on purpose, then it's fine as it is.

Comment (kortschak) [reply dsp/window/window.go]: Done.

**Comment by kortschak** [comment]
PTAL

File: dsp/window/window.go (hunk 3 above)
Comment (kortschak) [reply dsp/window/window.go]: I think you are right though. Having it explicit helps explain the intervals comment and makes it easier for reviewers to see errors.

**Pull request merged**